src/core/opamSystem.ml

```diff
@@ -175,6 +175,44 @@ let write file contents =
   output_string oc contents;
   close_out oc
 
+let setup_copy ?(chmod = fun x -> x) ~src ~dst () =
+  let ic = open_in_bin src in
+  let oc =
+    try
+      let perm =
+        (Unix.fstat (Unix.descr_of_in_channel ic)).st_perm |> chmod
+      in
+      open_out_gen
+        [ Open_wronly; Open_creat; Open_trunc; Open_binary ]
+        perm dst
+    with exn ->
+      close_in ic;
+      raise exn
+  in
+  (ic, oc)
+
+let copy_channels =
+  let buf_len = 65536 in
+  let buf = Bytes.create buf_len in
+  let rec loop ic oc =
+    match input ic buf 0 buf_len with
+    | 0 -> ()
+    | n ->
+      output oc buf 0 n;
+      loop ic oc
+  in
+  loop
+
+let copy_file_aux ?chmod ~src ~dst () =
+  let close_channels ic oc =
+    OpamStd.Exn.finally (fun () -> close_in ic) (fun () -> close_out oc) in
+  try
+    let ic, oc = setup_copy ?chmod ~src ~dst () in
+    OpamStd.Exn.finally (fun () -> close_channels ic oc)
```

Comment:
It could be nice to delete the partial destination file in case of error, but not sure.

Reply:
sounds reasonable.